Woodoku/src/Entity/WoodokuBoard.py

```diff
@@ -25,89 +28,171 @@ def remove_blocks(self, blocks_coord: List[Tuple[int, int]]) -> None:
+        can be added at `(x,y)`.
 
         This method includes actions:
-        1. Clear winning row/column/3x3 block
+        1. Clear winning row/column/3x3 box
         2. Call scoreAgent to calculate score
 
-        :param shape: The shape
-        :param x: x coordinate
-        :param y: y coordinate
-        :return: none
+        Args:
+            shape (WoodokuShape): The shape to be added
+            x (int): x coordinate
+            y (int): y coordinate
         """
-        pass
+        # add shape to block
+        shape_blocks = shape.map_to_board_at(x, y)
+        self.__representation.add_blocks(shape_blocks)
+
+        # determine groups and clear the groups
+        group_info, group_blocks = self.__find_groups()
+        self.__scoreAgent.calculate_winning(group_info)
+        self.__representation.remove_blocks(group_blocks)
 
     def get_score(self) -> int:
-        pass
+        return self.__scoreAgent.get_score()
 
-    def __find_winning_blocks(self) -> Dict[str, int]:
-        """
-        Check current board and see if there is any winning blocks such as complete rows, columns or 3x3 block.
-        :return: A dictionary mapping from name of the winning block to the number of winning block.
-            e.g. { "row": 1, "column": 2, "3x3 block": 0 }
-        """
-        pass
+    def __find_groups(self) -> Tuple[Dict[str, int], Set[Tuple[int, int]]]:
+        """Check current board and see if there is any groups such as
+        complete rows, columns or 3x3 box and report them. Remove all groups.
 
-    def __get_row_all_coords(self, row_num: int) -> List[Tuple[int, int]]:
-        pass
+        Terminology Reference:
+        [Wikipedia Sudoku Glossary](https://en.wikipedia.org/wiki/Glossary_of_Sudoku#Terminology_and_grid_layout)
 
-    def __get_column_all_coords(self, column_num: int) -> List[Tuple[int, int]]:
-        pass
+        Returns:
+            Tuple[Dict[str, int], Set[Tuple[int,int]]]:
+                1. A dictionary mapping from name of the group to the number of groups.
+                e.g. { "row": 1, "column": 2, "box": 0 }
+                2. Set of block coordinates for the group
 
-    def __get_3x3_all_coords(self, index: int) -> List[Tuple[int, int]]:
         """
-        The index of 3x3 block is as following:
-        _____________
-        | 0 | 1 | 2 |
-        | 3 | 4 | 5 |
-        | 6 | 7 | 8 |
-        _____________
-
-        :param index: the index as above
-        :return: All the coords in this 3x3 block.
+        info = {"row": 0, "column": 0, "box": 0}
+        rep = self.__representation
+        group_blocks = set()  # use set to handle overlapping group removal
+        for index in range(9):
+            row_blocks = self.__get_row_coords(index)
+            col_blocks = self.__get_col_coords(index)
+            box_blocks = self.__get_box_coords(index)
+
+            if rep.is_occupied(row_blocks):
```

Comment:
The definition for `_WoodokuBoardRepresentation.is_occupied` is unclear. If it returns true when **ANY** blocks are occpied, then the code here is incorrect.

Reply:
Will request and use`_WoodokuBoardRepresentation.is_all_occupied` to resolve the problem